test/run-drun/life-mut/life-v3.mo

```diff
@@ -0,0 +1,206 @@
+import P = "mo:⛔";
+import SM = "../stable-mem/StableMemory";
+
+actor Life {
+
+  object Random {
+    var state = 1;
+    public func next() : Bool {
+      state := (123138118391*state + 133489131) % 9999;
+      (state % 2 == 0)
+    };
+  };
+
+  class below(u : Nat) {
+    var i = 0;
+    public func next() : ?Nat { if (i >= u) null else {let j = i; i += 1; ?j} };
+  };
+
+  func readBitV2(bits : [var Nat64], index : Nat) : Bool {
+    let bit = P.natToNat64(index);
+    let mask : Nat64 = 1 << (bit % 64);
+    (bits[P.nat64ToNat(bit >> 6)] & mask) == mask
+  };
+
+  func readBit(offset : Nat32, index : Nat) : Bool {
+    let bit = P.natToNat32(index);
+    let mask : Nat32 = 1 << (bit % 32);
+    (SM.loadNat32(offset + ((bit >> 5)*4)) & mask) == mask
+  };
+
+  func writeBit(offset : Nat32, index : Nat, v : Bool) {
+    let bit = P.natToNat32(index);
+    let mask : Nat32 = 1 << (bit % 32);
+    let i = (bit >> 5)*4;
+    if v {
+      SM.storeNat32(offset + i, SM.loadNat32(offset + i) | mask)
+    }
+    else {
+      SM.storeNat32(offset + i, SM.loadNat32(offset + i) & ^mask)
+    };
+    assert (readBit(offset, index) == v);
+  };
+
+  type Cell = Bool;
+
+  type State = {
+    #v1 : [[var Cell]];
+    #v2 : {size : Nat; bits : [var Nat64]};
+    #v3 : {size : Nat; offset : Nat32}
+  };
```

Comment:
@crusso Just reading this carefully now.  Very interesting.  But, to sustain this pattern indefinitely, the (static) variant type's representation grows without bound, correct?  (with ever more variants)

That seems intuitively unavoidable to me, but does seem to create a tension with the desire not to grow an unbounded history of past representations within the current one (paraphrasing what I imagine @rossberg to mean when we talk about "not storing full history" in our stable memory story).

WDYT, out of curiosity?

Reply:
Actually, I believe you could remove old tags in the type just fine once all canisters running this code have been successfully upgraded.

In any case, this is linear in the _version_ history, which I assume is several orders of magnitude smaller than the call/data history that we need to worry about wrt scalability.

Reply:
> Actually, I believe you could remove old tags in the type just fine once all canisters running this code have been successfully upgraded.

I see, good point.  Notably, that's a little different than with a pubic API and our Candid upgrade story, where we must assume that some old clients always lurk out in the wild, right?

> In any case, this is linear in the version history, which I assume is several orders of magnitude smaller than the call/data history that we need to worry about wrt scalability.

Yes, absolutely.  My original thought a concern about the *concept* of growing "static history", not the practice of it, which I agree is not concerning.

Reply:
>Notably, that's a little different than with a pubic API and our Candid upgrade story, where we must assume that some old clients always lurk out in the wild, right?

Yes, exactly. For the public Candid interface, you generally do not know the set of all clients. For the private stable interface, your future self is the only client.